ark/utils/test_utils.py

```diff
@@ -0,0 +1,256 @@
+    mass_map = tuple(enumerate(channel_names, 1))
+
+    for i, fov in enumerate(fov_names):
+        tif_obj = mi.MibiImage(tif_data[i, :, :, :],
+                               mass_map,
+                               **MIBITIFF_METADATA)
+
+        tiffpath = os.path.join(base_dir, f'{fov}.tiff')
+        tiff.write(tiffpath, tif_obj, dtype=dtype)
+        filelocs[fov] = tiffpath
+
+    return filelocs, tif_data
+
+
+def _create_reverse_multitiff(base_dir, fov_names, channel_names, shape, sub_dir, fills, dtype):
+    tif_data = _gen_tif_data(len(channel_names), len(fov_names), shape, fills, dtype)
+
+    filelocs = {}
+
+    for i, fov in enumerate(fov_names):
+        tiffpath = os.path.join(base_dir, f'{fov}.tiff')
+        io.imsave(tiffpath, tif_data[:, :, :, i], plugin='tifffile')
+        filelocs[fov] = tiffpath
+
+    tif_data = np.swapaxes(tif_data, 0, -1)
+
+    return filelocs, tif_data
+
+
+def _create_labels(base_dir, fov_names, comp_names, shape, sub_dir, fills, dtype):
+    label_data = _gen_label_data(len(fov_names), len(comp_names), shape, dtype)
+
+    filelocs = {}
+
+    for i, fov in enumerate(fov_names):
+        tiffpath = os.path.join(base_dir, f'{fov}.tiff')
+        io.imsave(tiffpath, label_data[i, :, :, 0], plugin='tifffile')
+        filelocs[fov] = tiffpath
+
+    return filelocs, label_data
+
+
+TIFFMAKERS = {
+    'tiff': _create_tifs,
+    'multitiff': _create_multitiff,
+    'reverse_multitiff': _create_reverse_multitiff,
+    'mibitiff': _create_mibitiff,
+    'labels': _create_labels,
+}
+
+
+def create_paired_xarray_fovs(base_dir, fov_names, channel_names, img_shape=(1024, 1024),
+                              mode='tiff', delimiter=None, sub_dir=None, fills=False,
+                              dtype="int8"):
+
+    if not os.path.isdir(base_dir):
+        raise FileNotFoundError(f'{base_dir} is not a directory')
+
+    if fov_names is None or fov_names is []:
+        raise ValueError('No fov names were given...')
+
+    if channel_names is None or channel_names is []:
+        raise ValueError('No image names were given...')
+
+    if not isinstance(fov_names, list):
+        fov_names = [fov_names]
+
+    if not isinstance(channel_names, list):
+        channel_names = [channel_names]
+
+    filelocs, tif_data = TIFFMAKERS[mode](base_dir, fov_names, channel_names, img_shape, sub_dir,
+                                          fills, dtype)
+
+    if delimiter is not None:
+        fov_ids = [fov.split(delimiter)[0] for fov in fov_names]
+    else:
+        fov_ids = fov_names
+
+    if 'multitiff' in mode:
+        channel_names = range(len(channel_names))
+
+    if mode == 'labels':
+        data_xr = make_labels_xarray(tif_data, fov_ids, *img_shape, channel_names)
+    else:
+        data_xr = make_images_xarray(tif_data, fov_ids, *img_shape, channel_names)
+
+    return filelocs, data_xr
+
+
+def xarrays_are_equal(a_xr, b_xr, sortdims=("fovs", "channels")):
+    try:
+        if len(sortdims) > 1:
```

Comment:
What's the scenario where this comes up? Is it easier to sort here than to ensure that our test data is generated with the same order that the functions will create it in?

Reply:
the order that files are loaded in seems nondeterministic (but *usually* alphabetical), which makes the fovs column out of order; same issue comes up with channels for single channel tifs.  I figured it's better to do all sorting within the function.

Also, the reason `sortdims` is parametrized is because `label_xr`'s have `"components"` as their 4th dimension.  (although now that I think about it, I don't see why it can't just be `sortdims=(0, 3)`).  It will also take a list as input, but whatever it is it just needs to be iterable

Reply:
I think all of the data_loaders have a for loop that loops over the channels and FOVs list, either the provided one or everything in the folder with io_utils. So along as we've sorted the list that it's looping over it should be deterministic

Reply:
cool, I'll make sure that those are sorted in the data loaders.  That way, we can just use the regular `.equals` method on xarrays, instead of a custom one.